### DISCUSSION TITLE
[Serializer] Masking data?

### DISCUSSION THREAD

**Discussion by mpdude** [symfony/symfony#49661] [Q&A] [answered]
I am using the Serializer to serialize object graphs that ultimately will end up in log files or long-time storage. For data protection/security/compliance reasons, I need to mask sensitive data during this process.

In the most simple case, think of walking through the entire array of "normalized" data and doing `str_replace` for things like passwords hashes, secrets, credit card numbers etc. – before the encoder turns this data into the result string.

My question is: Is this use case already supported somehow, am I missing the relevant settings or code parts?

I can probably solve that by writing a decorating encoder that will do the processing before forwarding to another encoder.

However, it feels that this should be more of a normalizer or post-normalizer task than an encoding thing.

Does anyone else consider such a feature useful?

**Comment by ghost** [comment]
Maybe Serializer Component could introduce something similar to PHP's `#\SensitiveParameter` but for properties (and accesor methods).  
```php
class Foo 
{
  #[Serializer\SensitiveValue] // or SentitiveContent or SensitiveProperty or just Sensitive
  public string $field;
}
```

**Comment by lyrixx** [comment, marked as answer]
You can use [serialization groups](https://symfony.com/doc/current/serializer.html#using-serialization-groups-attributes). With groups you can control exactly what you serialize. This is clearly the way to do.

I suggest you to name the group with the following pattern `project:read`, `project:read:details`, etc

I made [a presentation](https://speakerdeck.com/lyrixx/symfony-live-2017-serializer) about the serializer few years ago. You can read it if the doc is not enough.

**Reply by mpdude** [reply]
Re-reading my use case description, I realize it was not precise enough.

What I want to do is to suppress particular _data_ in the fields. So, for example, put _all_ fields from the given object tree into the serialize result. But, before the final result is encoded, walk over all (normalized) data and perform string replacements of sensitive data.

Does that make sense?

**Reply by lyrixx** [reply]
This is a bit weird because it means  data are not really cleaned / designed 

But you can use special getter and do the replacement in each getter

**Reply by mpdude** [reply]
Ok, so I will have to think this through. 

Thank you for your help and suggestions.
